src/nal/sps.rs

```diff
@@ -292,33 +295,66 @@ impl From<ProfileIdc> for u8 {
     }
 }
 
-pub struct ScalingList {
-    // TODO
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ScalingList<const S: usize> {
+    NotPresent,
+    UseDefault,
+    List([NonZeroU8; S]),
 }
-impl ScalingList {
-    pub fn read<R: BitRead>(r: &mut R, size: u8) -> Result<ScalingList, ScalingMatrixError> {
-        let mut scaling_list = vec![];
-        let mut last_scale = 8;
-        let mut next_scale = 8;
-        let mut _use_default_scaling_matrix_flag = false;
-        for j in 0..size {
-            if next_scale != 0 {
-                let delta_scale = r.read_se("delta_scale")?;
-                if delta_scale < -128 || delta_scale > 127 {
-                    return Err(ScalingMatrixError::DeltaScaleOutOfRange(delta_scale));
-                }
-                next_scale = (last_scale + delta_scale + 256) % 256;
-                _use_default_scaling_matrix_flag = j == 0 && next_scale == 0;
+
+/// returns 'use_default_scaling_matrix_flag'
+fn fill_scaling_list<R: BitRead>(
+    r: &mut R,
+    scaling_list: &mut [NonZeroU8],
+) -> Result<bool, ScalingMatrixError> {
+    let mut last_scale = 8;
+    let mut next_scale = 8;
+    let mut use_default_scaling_matrix_flag = false;
+
+    for j in 0..scaling_list.len() {
+        if next_scale != 0 {
+            let delta_scale = r.read_se("delta_scale")?;
+            if delta_scale < -128 || delta_scale > 127 {
+                return Err(ScalingMatrixError::DeltaScaleOutOfRange(delta_scale));
             }
-            let new_value = if next_scale == 0 {
-                last_scale
-            } else {
-                next_scale
-            };
-            scaling_list.push(new_value);
-            last_scale = new_value;
+            next_scale = (last_scale + delta_scale + 256) % 256;
+            use_default_scaling_matrix_flag = j == 0 && next_scale == 0;
+        }
+        let new_value = if next_scale == 0 {
+            last_scale
+        } else {
+            next_scale
+        };
+        // unwrap is ok here, because:
```

Comment:
I agree with the argument in this comment, but I think it's possible to avoid the `.unwrap()` altogether:

* keep `last_scale` as a `NonZeroU8`, starting from `const { NonZeroU8::new(8).unwrap() }` (msrp 1.83, the alternative of `match { ..., None => panic!() }` works a few versions before)
* update via `new_value = NonZeroU8::new(next_scale as u8).unwrap_or(last_value)`